pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+		if amount > 0 {
+			for _, strongTip := range t.strongTipSet.RandomUniqueEntries(amount + seenStrongTips.Size()) {
+				if seenStrongTips.Add(strongTip.Block().ID()) {
+					if strongTips = append(strongTips, strongTip); len(strongTips) == amount {
+						return strongTips
+					}
+				}
+			}
+		}
+
+		return strongTips
+	}
+
+	_ = t.conflictDAG.ReadConsistent(func(conflictDAG conflictdag.ReadLockedConflictDAG[iotago.TransactionID, iotago.OutputID, booker.BlockVotePower]) error {
+		likedConflicts := advancedset.New[iotago.TransactionID]()
+
+		likedInsteadReferences := func(tipMetadata *TipMetadata) (references []iotago.BlockID, updatedLikedConflicts *advancedset.AdvancedSet[iotago.TransactionID], err error) {
+			necessaryReferences := make(map[iotago.TransactionID]iotago.BlockID)
+			if err = conflictDAG.LikedInstead(tipMetadata.Block().ConflictIDs()).ForEach(func(likedConflictID iotago.TransactionID) error {
+				transactionMetadata, exists := t.memPool.TransactionMetadata(likedConflictID)
+				if !exists {
+					return xerrors.Errorf("transaction required for liked instead reference (%s) not found in mem-pool", likedConflictID)
+				}
+
+				necessaryReferences[likedConflictID] = lo.First(transactionMetadata.Attachments())
+
+				return nil
+			}); err != nil {
+				return nil, nil, err
+			}
+
+			references, updatedLikedConflicts = make([]iotago.BlockID, 0), likedConflicts.Clone()
+			for conflictID, attachmentID := range necessaryReferences {
+				if updatedLikedConflicts.Add(conflictID) {
+					references = append(references, attachmentID)
+				}
+			}
+
+			if len(references) > t.optMaxLikedInsteadReferencesPerParent {
+				return nil, nil, xerrors.Errorf("too many liked instead references (%d) for block %s", len(references), tipMetadata.Block().ID())
+			}
+
+			return references, updatedLikedConflicts, nil
+		}
+
+		for strongTipCandidates := selectStrongTips(amount - len(references[model.StrongParentType])); len(strongTipCandidates) != 0; strongTipCandidates = selectStrongTips(amount - len(references[model.StrongParentType])) {
+			for _, strongTip := range strongTipCandidates {
+				if addedLikedInsteadReferences, updatedLikedConflicts, err := likedInsteadReferences(strongTip); err != nil {
+					// TODO: LOG REASON FOR DOWNGRADE?
+
+					strongTip.setTipPool(t.determineTipPool(strongTip, tipmanager.WeakTipPool))
+				} else if len(addedLikedInsteadReferences) <= t.optMaxLikedInsteadReferences-len(references[model.ShallowLikeParentType]) {
+					references[model.StrongParentType] = append(references[model.StrongParentType], strongTip.Block().ID())
+					references[model.ShallowLikeParentType] = append(references[model.ShallowLikeParentType], addedLikedInsteadReferences...)
+
+					likedConflicts = updatedLikedConflicts
+				}
+			}
+		}
+
+		references[model.WeakParentType] = lo.Map(t.weakTipSet.RandomUniqueEntries(t.optMaxWeakReferences), func(sourceType *TipMetadata) iotago.BlockID {
```

Comment:
We should also possibly fix opinion of WeakParents as well. It could happen that a tip that the payload of the tip is not liked or even rejected and we don't want to cast such vote. This could also result in issuing subjectively invalid blocks. 
If we cannot fix the vote, then we mark the tip as Dropped. Currently we only drop tips when downgrading from Strong to Weak, but if a block ends up in the Weak pool then it can never be Dropped.

Reply:
I think this is a valid concern - we should re-evaluate the membership in the weak tip set upon selection and refrain from rescuing things that might no longer be liked, which would automatically cause eviction from the weak tip pool.

I will make the necessary adjustments.

Reply:
I fixed it.